src/TemplateCore.php

```diff
@@ -175,4 +182,50 @@ protected function endSection() : void
     }
 
     abstract protected function render(string $__NAME__, array $__VARS__ = []) : string;
+
+    protected function pushRenderStack(string $name) : string
+    {
+        $prefix = empty($this->renderStack)
+            ? ''
+            : dirname(end($this->renderStack));
+
+        $orig = $name;
+        $name = $this->resolveDots($name, $prefix);
+
+        if (strpos($name, '..') !== false) {
+            throw new Exception\TemplateNotFound(
+                PHP_EOL . "Could not resolve dots in template name." .
+                PHP_EOL . "Original name: '{$orig}'" .
+                PHP_EOL . "Resolved into: '{$name}'" .
+                PHP_EOL . "Probably too many '../' in the original name."
+            );
+        }
+
+        array_push($this->renderStack, $name);
+        return $name;
+    }
+
+    protected function resolveDots(string $name, string $prefix) : string
+    {
+        $name = ltrim(trim($name), '/');
+
+        if (substr($name, 0, 2) === './') {
```

Comment:
https://www.php.net/manual/en/function.str-starts-with.php (PHP8)

```suggestion
        if (str_starts_with($name, './')) {
```